src/stages/stage3_events.c

```diff
@@ -14,1666 +14,1005 @@
+	INVOKE_TASK(corner_fairy, p4, p4 - offs + offs*I, p1 - offs - offs*I);
 
-		PROJECTILE(
-			.proto = (frand() > 0.5)? pp_thickrice : pp_rice,
-			.pos = p->pos,
-			.color = RGB(0.3, 0.7 + 0.3 * psin(a/3.0 + t/20.0), 0.3),
-			.rule = accelerated,
-			.args = {
-				0,
-				0.005*cexp(I*(M_PI*2 * sin(a/5.0 + t/20.0))),
-			},
-		);
+	WAIT(90);
 
-		play_sound("redirect");
+	if(global.diff > D_Normal) {
+		INVOKE_TASK(corner_fairy, p1, p1 - offs*I, p2 + offs);
+		INVOKE_TASK(corner_fairy, p2, p2 + offs, p3 + offs*I);
+		INVOKE_TASK(corner_fairy, p3, p3 + offs*I, p4 - offs);
+		INVOKE_TASK(corner_fairy, p4, p4 - offs, p1 + offs*I);
 	}
-
-	return result;
 }
 
-static int scuttle_lethbite_proj(Projectile *p, int time) {
-	if(time < 0) {
-		return ACTION_ACK;
-	}
-
-	#define A0_PROJ_START 120
-	#define A0_PROJ_CHARGE 20
-	TIMER(&time)
-
-	FROM_TO(A0_PROJ_START, A0_PROJ_START + A0_PROJ_CHARGE, 1)
-		return 1;
+/*
+ * Bosses
+ */
 
-	AT(A0_PROJ_START + A0_PROJ_CHARGE + 1) if(p->type != PROJ_DEAD) {
-		p->args[1] = 3;
-		p->args[0] = (3 + 2 * global.diff / (float)D_Lunatic) * cexp(I*carg(global.plr.pos - p->pos));
+// scuttle
 
-		int cnt = 3, i;
-		for(i = 0; i < cnt; ++i) {
-			tsrand_fill(2);
+TASK_WITH_INTERFACE(scuttle_intro, BossAttack) {
+	Boss *boss = INIT_BOSS_ATTACK();
+	BEGIN_BOSS_ATTACK();
 
-			PARTICLE(
-				.sprite = "smoothdot",
-				.color = RGBA(0.8, 0.6, 0.6, 0),
-				.draw_rule = Shrink,
-				.rule = enemy_flare,
-				.timeout = 100,
-				.args = {
-					cexp(I*(M_PI*anfrand(0))) * (1 + afrand(1)),
-					add_ref(p)
-				},
-			);
+	boss->move = move_towards(VIEWPORT_W/2.0 + 100.0*I, 0.04);
+}
 
-			float offset = global.frames/15.0;
-			if(global.diff > D_Hard && global.boss) {
-				offset = M_PI+carg(global.plr.pos-global.boss->pos);
-			}
+TASK_WITH_INTERFACE(scuttle_outro, BossAttack) {
+	Boss *boss = INIT_BOSS_ATTACK();
+	BEGIN_BOSS_ATTACK();
 
-			PROJECTILE(
-				.proto = pp_thickrice,
-				.pos = p->pos,
-				.color = RGB(0.4, 0.3, 1.0),
-				.rule = linear,
-				.args = {
-					-cexp(I*(i*2*M_PI/cnt + offset)) * (1.0 + (global.diff > D_Normal))
-				},
-			);
-		}
+	boss->move = move_towards(VIEWPORT_W/2 - 200.0*I, 0.05);
+}
 
-		play_sound("redirect");
-		play_sound("shot1");
-		spawn_projectile_highlight_effect(p);
-	}
+TASK(scuttle_delaymove, { BoxedBoss boss } ) {
+	Boss *boss = ENT_UNBOX(ARGS.boss);
 
-	return asymptotic(p, time);
-	#undef A0_PROJ_START
-	#undef A0_PROJ_CHARGE
+	WAIT(100);
+	boss->move.attraction_point = VIEWPORT_W/2+VIEWPORT_W/3;
+	boss->move.acceleration = 0.08;
 }
 
-static void scuttle_lethbite(Boss *boss, int time) {
-	int i;
-	TIMER(&time)
+TASK_WITH_INTERFACE(scuttle_lethbite, BossAttack) {
+	Boss *boss = INIT_BOSS_ATTACK();
+	BEGIN_BOSS_ATTACK();
 
-	GO_TO(boss, VIEWPORT_W/2+VIEWPORT_W/3*sin(time/300) + I*cimag(boss->pos), 0.01)
+	int intensity = difficulty_value(18, 19, 20, 21);
+	INVOKE_TASK_DELAYED(300, scuttle_delaymove, ENT_BOX(boss));
```

Comment:
Should be a subtask; that way it'll be cancelled if the attack finishes too early.